File: ResumeCore/iosApp/iosApp/Util/PeriodFormatter.swift
import Foundation
import Shared

private let enMonths = [
    "Jan", "Feb", "Mar", "Apr", "May", "Jun",
    "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"
]

enum PeriodFormatter {

    static func format(_ period: Period) -> String {
        let useChinese = L10n.isChinese
        let start = period.start
        let end = period.end
        let startsInJan = start.month == 1
        let endsInDec = end?.month == 12
        if startsInJan && (endsInDec || end == nil) {
            if let end {
                return start.year == end.year
                    ? "\(start.year)"
                    : "\(start.year) ─ \(end.year)"
            }
            return "\(start.year)"
        }
        let endLabel = end.map { yearMonth($0, useChinese: useChinese) } ?? L10n.present
        return "\(yearMonth(start, useChinese: useChinese)) ─ \(endLabel)"
    }

    static func duration(_ period: Period) -> String {
        let endYM = period.end ?? nowYearMonth()
        let months = max(0, monthsBetween(period.start, endYM))
        let y = months / 12
        let m = months % 12
        switch (y, m) {
        case (0, _): return L10n.durationMonths(m)
        case (_, 0): return L10n.durationYears(y)
        default:     return L10n.durationYearsMonths(y, m)
        }
    }

    static func yearRangeShort(_ period: Period) -> String {
        if let end = period.end, end.year != period.start.year {
            return "\(period.start.year)–\(end.year)"
        }
        return "\(period.start.year)"
    }

    static func totalYearsApprox(_ periods: [Period]) -> String {
        let totalMonths = periods.map { months in
            let end = months.end ?? nowYearMonth()
            return max(0, monthsBetween(months.start, end))
        }.reduce(0, +)
        let years = totalMonths / 12
        return "\(years)+"
    }

    private static func yearMonth(_ ym: YearMonth, useChinese: Bool) -> String {
        if useChinese {
            let mm = String(format: "%02d", ym.month)
            return "\(ym.year)/\(mm)"
        }
        let idx = Int(ym.month) - 1
        let name = (idx >= 0 && idx < 12) ? enMonths[idx] : "?"
        return "\(name) \(ym.year)"
    }

    private static func monthsBetween(_ a: YearMonth, _ b: YearMonth) -> Int {
        Int((b.year - a.year) * 12 + (b.month - a.month))
    }

    private static func nowYearMonth() -> YearMonth {
        let cal = Calendar(identifier: .gregorian)
        let comps = cal.dateComponents([.year, .month], from: Date())
        return YearMonth(year: Int32(comps.year ?? 2026), month: Int32(comps.month ?? 1))
    }
}

File: ResumeCore/composeApp/src/androidMain/kotlin/com/danielchi0716/resume/core/ui/format/PeriodFormatter.kt
package com.danielchi0716.resume.core.ui.format

import androidx.compose.runtime.Composable
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.res.stringResource
import com.danielchi0716.resume.core.R
import com.danielchi0716.resume.core.model.Period
import com.danielchi0716.resume.core.model.YearMonth

private val EnMonthShort = listOf(
    "Jan", "Feb", "Mar", "Apr", "May", "Jun",
    "Jul", "Aug", "Sep", "Oct", "Nov", "Dec",
)

@Composable
fun isChineseLocale(): Boolean {
    val cfg = LocalConfiguration.current
    return cfg.locales[0].language.equals("zh", ignoreCase = true)
}

@Composable
fun formatPeriod(period: Period): String {
    val useChinese = isChineseLocale()
    val present = stringResource(R.string.period_present)
    return formatPeriodInternal(period, useChinese, present)
}

@Composable
fun formatDuration(period: Period): String {
    val end = period.end ?: nowYearMonth()
    val months = monthsBetween(period.start, end).coerceAtLeast(0)
    val y = months / 12
    val m = months % 12
    return when {
        y == 0 -> stringResource(R.string.duration_months, m)
        m == 0 -> stringResource(R.string.duration_years, y)
        else -> stringResource(R.string.duration_years_months, y, m)
    }
}

fun yearRangeShort(period: Period): String {
    val end = period.end
    return if (end == null || period.start.year == end.year) {
        "${period.start.year}"
    } else {
        "${period.start.year}–${end.year}"
    }
}

private fun formatPeriodInternal(
    period: Period,
    useChinese: Boolean,
    presentLabel: String,
): String {
    val start = period.start
    val end = period.end
    val startsInJan = start.month == 1
    val endsInDec = end != null && end.month == 12
    if (startsInJan && (endsInDec || end == null)) {
        return when {
            end == null || start.year == end.year -> "${start.year}"
            else -> "${start.year} ─ ${end.year}"
        }
    }
    val endLabel = end?.let { formatYearMonth(it, useChinese) } ?: presentLabel
    return "${formatYearMonth(start, useChinese)} ─ $endLabel"
}

private fun formatYearMonth(ym: YearMonth, useChinese: Boolean): String =
    if (useChinese) "${ym.year}/${ym.month.toString().padStart(2, '0')}"
    else "${EnMonthShort[ym.month - 1]} ${ym.year}"

private fun monthsBetween(start: YearMonth, end: YearMonth): Int =
    (end.year - start.year) * 12 + (end.month - start.month)

internal fun nowYearMonth(): YearMonth {
    val cal = java.util.Calendar.getInstance()
    return YearMonth(
        cal.get(java.util.Calendar.YEAR),
        cal.get(java.util.Calendar.MONTH) + 1,
    )
}

internal fun totalMonthsAtNow(period: Period): Int {
    val end = period.end ?: nowYearMonth()
    return monthsBetween(period.start, end).coerceAtLeast(0)
}
